src/io/flutter/coverage/ScriptManager.java

```diff
@@ -0,0 +1,191 @@
+ * found in the LICENSE file.
+ */
+package io.flutter.coverage;
+
+import com.intellij.openapi.util.Pair;
+import com.intellij.openapi.util.Ref;
+import com.intellij.openapi.vfs.VirtualFile;
+import com.intellij.openapi.vfs.VirtualFileManager;
+import com.intellij.util.concurrency.Semaphore;
+import gnu.trove.THashMap;
+import gnu.trove.TIntObjectHashMap;
+import org.dartlang.vm.service.VmService;
+import org.dartlang.vm.service.consumer.GetIsolateConsumer;
+import org.dartlang.vm.service.consumer.GetLibraryConsumer;
+import org.dartlang.vm.service.consumer.GetObjectConsumer;
+import org.dartlang.vm.service.element.*;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+class ScriptManager {
+  private static final long RESPONSE_WAIT_TIMEOUT = 3000;
+
+  @NotNull private final VmService vmService;
+
+  private final VirtualFileManager virtualFileManager = VirtualFileManager.getInstance();
+
+  private IsolateRef isolateRef;
+  private final Map<String, Script> scriptMap = new HashMap<>();
+  private final Map<String, TIntObjectHashMap<Pair<Integer, Integer>>> linesAndColumnsMap = new THashMap<>();
+
+  public ScriptManager(@NotNull VmService vmService) {
+    this.vmService = vmService;
+  }
+
+  public void reset() {
+    scriptMap.clear();
+  }
+
+  public void setCurrentIsolate(IsolateRef isolateRef) {
+    this.isolateRef = isolateRef;
+  }
+
+  @Nullable
+  private Isolate getCurrentIsolate() {
+    final Ref<Isolate> resultRef = Ref.create();
+    final Semaphore semaphore = new Semaphore();
+    semaphore.down();
+
+    vmService.getIsolate(isolateRef.getId(), new GetIsolateConsumer() {
+      @Override
+      public void received(Isolate isolate) {
+        resultRef.set(isolate);
+        semaphore.up();
+      }
+
+      @Override
+      public void received(Sentinel sentinel) {
+        semaphore.up();
+      }
+
+      @Override
+      public void onError(RPCError error) {
+        semaphore.up();
+      }
+    });
+    semaphore.waitFor(RESPONSE_WAIT_TIMEOUT);
+    return resultRef.get();
+  }
+
+  @Nullable
+  public ScriptRef getScriptRefFor(@NotNull VirtualFile file) {
+    final Isolate isolate = getCurrentIsolate();
+    if (isolate == null) {
+      return null;
+    }
+
+    for (LibraryRef libraryRef : isolate.getLibraries()) {
+      final String uri = libraryRef.getUri();
+
+      if (uri.startsWith("file:")) {
+        final VirtualFile libraryFile = virtualFileManager.findFileByUrl(uri);
+
+        if (file.equals(libraryFile)) {
+          final Library library = getLibrary(libraryRef);
+          if (library != null) {
+            if (!library.getScripts().isEmpty()) {
+              return library.getScripts().get(0);
+            }
+          }
+        }
+      }
+    }
+
+    return null;
+  }
+
+  @Nullable
+  private Library getLibrary(LibraryRef libraryRef) {
```

Comment:
Consider using
 CompletableFuture<Library> getLibrary(LibraryRef instance)
from EvalOnDartLibrary also performs this task. Perhaps add a TODO to standardize to a single implementation going forward.